File: backend/app/api/v1/transactions.py
import uuid
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.visibility import VisibilityContext, get_visibility_ctx
from app.db.base import get_session
from app.schemas.transaction import (
    BulkCategorizeRequest,
    PaginatedTransactions,
    TransactionCreate,
    TransactionResponse,
    TransactionUpdate,
)
from app.services.transaction import TransactionService

router = APIRouter()


@router.get("/accounts/{account_id}/transactions", response_model=PaginatedTransactions)
async def list_transactions(
    account_id: uuid.UUID,
    from_: date | None = Query(None, alias="from"),
    to: date | None = Query(None),
    category_id: uuid.UUID | None = Query(None),
    is_reviewed: bool | None = Query(None),
    is_transfer: bool | None = Query(None),
    real_estate_property_id: uuid.UUID | None = Query(None),
    search: str | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=500),
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> PaginatedTransactions:
    svc = TransactionService(session)
    items, total = await svc.list_for_account(
        ctx,
        account_id,
        from_date=from_,
        to_date=to,
        category_id=category_id,
        is_reviewed=is_reviewed,
        is_transfer=is_transfer,
        real_estate_property_id=real_estate_property_id,
        search=search,
        page=page,
        page_size=page_size,
    )
    return PaginatedTransactions(
        items=[TransactionResponse.model_validate(t) for t in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post(
    "/accounts/{account_id}/transactions", response_model=TransactionResponse, status_code=201
)
async def create_transaction(
    account_id: uuid.UUID,
    data: TransactionCreate,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> TransactionResponse:
    svc = TransactionService(session)
    transaction = await svc.create(ctx, account_id, data)
    await session.commit()
    await session.refresh(transaction)
    return TransactionResponse.model_validate(transaction)


@router.patch(
    "/accounts/{account_id}/transactions/bulk-categorize",
    response_model=list[TransactionResponse],
)
async def bulk_categorize(
    account_id: uuid.UUID,
    data: BulkCategorizeRequest,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> list[TransactionResponse]:
    svc = TransactionService(session)
    transactions = await svc.bulk_categorize(ctx, account_id, data)
    await session.commit()
    return [TransactionResponse.model_validate(t) for t in transactions]


@router.get("/transactions/{transaction_id}", response_model=TransactionResponse)
async def get_transaction(
    transaction_id: uuid.UUID,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> TransactionResponse:
    svc = TransactionService(session)
    transaction = await svc.get(ctx, transaction_id)
    return TransactionResponse.model_validate(transaction)


@router.patch("/transactions/{transaction_id}", response_model=TransactionResponse)
async def update_transaction(
    transaction_id: uuid.UUID,
    data: TransactionUpdate,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> TransactionResponse:
    svc = TransactionService(session)
    transaction = await svc.update(ctx, transaction_id, data)
    await session.commit()
    await session.refresh(transaction)
    return TransactionResponse.model_validate(transaction)


@router.delete("/transactions/{transaction_id}", status_code=204)
async def delete_transaction(
    transaction_id: uuid.UUID,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> None:
    svc = TransactionService(session)
    await svc.delete(ctx, transaction_id)
    await session.commit()

File: frontend/src/api/transactions.ts
import { api } from "./client"
import type { PaginatedTransactions, TransactionCreate, TransactionResponse } from "./types"

export interface TransactionFilters {
  from?: string
  to?: string
  category_id?: string
  is_reviewed?: boolean
  is_transfer?: boolean
  real_estate_property_id?: string
  search?: string
  page?: number
  page_size?: number
}

function buildQuery(filters: TransactionFilters): string {
  const params = new URLSearchParams()
  if (filters.from) params.set("from", filters.from)
  if (filters.to) params.set("to", filters.to)
  if (filters.category_id) params.set("category_id", filters.category_id)
  if (filters.is_reviewed !== undefined) params.set("is_reviewed", String(filters.is_reviewed))
  if (filters.is_transfer !== undefined) params.set("is_transfer", String(filters.is_transfer))
  if (filters.real_estate_property_id) {
    params.set("real_estate_property_id", filters.real_estate_property_id)
  }
  if (filters.search) params.set("search", filters.search)
  if (filters.page) params.set("page", String(filters.page))
  if (filters.page_size) params.set("page_size", String(filters.page_size))
  const qs = params.toString()
  return qs ? `?${qs}` : ""
}

export const transactionsApi = {
  list: (accountId: string, filters: TransactionFilters = {}) =>
    api.get<PaginatedTransactions>(`/accounts/${accountId}/transactions${buildQuery(filters)}`),

  create: (accountId: string, data: TransactionCreate) =>
    api.post<TransactionResponse>(`/accounts/${accountId}/transactions`, data),

  update: (
    id: string,
    data: Partial<{
      transaction_date: string
      amount: string
      payee_normalized: string
      memo: string | null
      category_id: string | null
      is_transfer: boolean
      real_estate_property_id: string | null
      is_reviewed: boolean
    }>,
  ) => api.patch<TransactionResponse>(`/transactions/${id}`, data),

  delete: (id: string) => api.delete(`/transactions/${id}`),

  bulkCategorize: (accountId: string, transactionIds: string[], categoryId: string) =>
    api.patch<TransactionResponse[]>(`/accounts/${accountId}/transactions/bulk-categorize`, {
      transaction_ids: transactionIds,
      category_id: categoryId,
    }),
}
